### PR TITLE
Fix WinHttpHandler proxy discovery behavior

### LINKED ISSUE #6011: HttpClient does not support proxy PAC file with file:// scheme
I have a vanilla ASP.net vnext project, and when switching from `dnx451 `to `dnxcore50`, `HttpClient `stops working assuming an URI with an invalid protocol. The code itself is trivial:

        public IActionResult Get()
        {
            string baseUrl = "http://someserver.com/foo/";
            using (var client = new HttpClient() { BaseAddress = new Uri(baseUrl) })
            {
                var result = client.GetAsync("items").Result;
                ...
            }
        }

I tried importing different versions of `System.Net.Http ` (e.g. _"System.Net.Http": "4.0.1-beta-23409"_), including the latest stable and latest beta, but they are all the same when running under the CoreClr.

I didn't find another implementation of HttpClient for the CoreClr, but then, I have a hard time believing that something as elementary would be missing or broken. Am I missing something or using the wrong library?

Thanks,
Philipp

System.AggregateException: One or more errors occurred. ---> System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.Http.WinHttpException: The URL does not use a recognized protocol
   at System.Net.Http.WinInetProxyHelper.GetProxyForUrl(SafeWinHttpHandle sessionHandle, Uri uri, WINHTTP_PROXY_INFO& proxyInfo)
   at System.Net.Http.WinHttpHandler.SetRequestHandleProxyOptions(WinHttpRequestState state)
   at System.Net.Http.WinHttpHandler.SetRequestHandleOptions(WinHttpRequestState state)
   at System.Net.Http.WinHttpHandler.<StartRequest>d__103.MoveNext()
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at OrderManager.Controllers.JobsController.Get()
---> (Inner Exception #0) System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.Http.WinHttpException: The URL does not use a recognized protocol
   at System.Net.Http.WinInetProxyHelper.GetProxyForUrl(SafeWinHttpHandle sessionHandle, Uri uri, WINHTTP_PROXY_INFO& proxyInfo)
   at System.Net.Http.WinHttpHandler.SetRequestHandleProxyOptions(WinHttpRequestState state)
   at System.Net.Http.WinHttpHandler.SetRequestHandleOptions(WinHttpRequestState state)
   at System.Net.Http.WinHttpHandler.<StartRequest>d__103.MoveNext()
   --- End of inner exception stack trace --

### PR DESCRIPTION
**Pull request by davidsh**
In order to support WindowsProxyUsePolicy.UseWinInetProxy, WinHTTP provides the WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY option. This allows for reading the per-user registry settings to determine how to find a proxy for a given url. This includes using WPAD protocol to find and execute a Javascript PAC file. The WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY only works on Windows 8.1 and above. For Windows 7 and Windows 8, we have to manually do the work and the WinInetProxyHelper class provides for that by calling various WinHTTP APIs.

This fix is about having the AUTOMATIC_PROXY behaviors work the same whether done automatically by WinHTTP or manually using the WinInetProxyHelper class. In general, errors processing the auto discovery logic should be ignored and the proxy settings should fallback to manual settings if present in the registry.

This is difficult to test in an Xunit test since registry settings need to be changed as well as recycling the "WinHTTP Web Proxy Autodiscover Discovery Service" since it caches PAC file processing. So, this was tested manually for now and additional traces were added to the WinInetProxyHelper class.

Fixes #6011

### CODE REVIEW

**Comment by davidsh** [comment]
@CIPop @stephentoub PTAL

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 1]
```diff
@@ -25,16 +27,25 @@ public WinInetProxyHelper()
                     AutoDetect = proxyConfig.AutoDetect;
                     Proxy = Marshal.PtrToStringUni(proxyConfig.Proxy);
                     ProxyBypass = Marshal.PtrToStringUni(proxyConfig.ProxyBypass);
+
+                    WinHttpTraceHelper.Trace(
+                        "WinInetProxyHelper.ctor: AutoConfigUrl={0}, AutoDetect={1}, Proxy={2}, ProxyBypass={3}",
+                        AutoConfigUrl,
+                        AutoDetect,
+                        Proxy,
+                        ProxyBypass);
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: All of these arguments appear to be of value types (bool and IntPtr).  As such, this call is going to result in four allocations, even when tracing is disabled.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 2]
```diff
@@ -25,16 +27,25 @@ public WinInetProxyHelper()
...
-                    if (lastError != Interop.WinHttp.ERROR_FILE_NOT_FOUND)
-                    {
-                        throw WinHttpException.CreateExceptionUsingError(lastError);
-                    }
+                    // We match behavior of WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY and ignore errors.
+                    int lastError = Marshal.GetLastWin32Error();
+                    WinHttpTraceHelper.Trace("WinInetProxyHelper.ctor: error={0}", lastError);
                 }
+
+                WinHttpTraceHelper.Trace("WinInetProxyHelper.ctor: _useProxy={0}", _useProxy);
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Some for both of these Trace calls... lastError and _useProxy will both get boxed.  I'll stop commenting on these.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 3]
```diff
@@ -13,6 +13,8 @@ namespace System.Net.Http
     // is not supported (i.e. before Win8.1/Win2K12R2) in the WinHttpOpen() function.
     internal class WinInetProxyHelper
     {
+        bool _useProxy = false;
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: private

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 4]
```diff
@@ -92,50 +118,65 @@ public void GetProxyForUrl(SafeWinHttpHandle sessionHandle, Uri uri, out Interop
...
                 }
             } while (repeat);
+
+            // Fall back to manual settings if available.
+            if (!useProxy && !string.IsNullOrEmpty(Proxy))
+            {
+                proxyInfo.AccessType = Interop.WinHttp.WINHTTP_ACCESS_TYPE_NAMED_PROXY;
+                proxyInfo.Proxy = Marshal.StringToHGlobalUni(Proxy);
+                proxyInfo.ProxyBypass = string.IsNullOrEmpty(ProxyBypass) ?
+                    IntPtr.Zero : Marshal.StringToHGlobalUni(ProxyBypass);
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Where do these native strings get freed?

File: src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs [hunk 5]
```diff
@@ -751,7 +751,7 @@ public void SendAsync_NoAutomaticProxySupportAndUseWinInetSettingsWithAutoDetect
                 });
 
             Assert.Equal(Interop.WinHttp.WINHTTP_ACCESS_TYPE_NO_PROXY, APICallHistory.SessionProxySettings.AccessType);
-            Assert.Equal(Interop.WinHttp.WINHTTP_ACCESS_TYPE_NO_PROXY, APICallHistory.RequestProxySettings.AccessType);
+            Assert.Equal(null, APICallHistory.RequestProxySettings.AccessType);
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs]: Assert.Null

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 4 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: This code was always there, but just moved around in this method.

The caller, `WinHttpHandler.SetRequestHandleProxyOptions()` frees those native strings.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 3 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: fixed

File: src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs (hunk 5 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs]: fixed

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 1 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Actually, for this call they are: `string`, `bool`, `string`, `string`.  The `IntPtr` are in the proxyConfig struct. But they are converted to .NET strings. I'll add if I can add just specific overloads for these for now.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 2 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: I"ll add specific overloads for now.

Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Ok.  You could also make them generic methods.

Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: I plan on switching this tracing stuff to EventSource sometime soon. So, hopefully, there will be an easier/cleaner coding pattern.

**Comment by stephentoub** [comment]
LGTM

**Pull request merged**